enclave/core/sgx/once.c

```diff
@@ -5,34 +5,39 @@
 #include <openenclave/internal/thread.h>
 #include <openenclave/internal/utils.h>
 
+#define SPINLOCK_NOT_USED 0
+#define SPINLOCK_ACTIVELY_HELD 1
+#define SPINLOCK_USED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
     OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
-    if (o == 0)
+    if (*once != SPINLOCK_USED)
     {
-        static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
-
-        oe_spin_lock(&_lock);
-
-        if (*once == 0)
+        oe_once_t retval = __sync_val_compare_and_swap(
```

Comment:
"Multiple threads could reach here simultaneously after checking whether the function has been invoked or not.
Only one of them must now invoke the function and others must wait for the function invocation to complete.

To determine who gets to invoke the function, each thread atomically compares the value of once to FUNC_NOT_INVOKED and if equal, sets the value to FUNC_BEING_INVOKED to signal to other threads that the function is being invoked. There return value of __sync_val_compare_and_swap determines which thread takes ownership.
If __sync_val_compare_and_swap returns FUNC_NOT_INVOKED, then that means this thread successfully set once to FUNC_BEING_INVOKED, and can now safely call func.
If __sync_val_compare_and_swap returns FUNC_BEING_INVOKED, this means another thread's __sync_val_compare_and_swap succeeded first and therefore this thread now has to wait for the other thread to complete (ie wait for once to become FUNC_INVOKED.
If __sync_val_compare_and_swap returns FUNC_INVOKED, that means another thread has already called the function and marked the once as complete. This thread can safely proceed."